docs/HYPERVISORS.md

```diff
@@ -0,0 +1,85 @@
+# EVE's use of hypervisors and hardware assisted virtualization
+
+EVE is slightly unusual in how it looks at hypervisors and hardware assisted virtualization technology. Unlike some of the other systems that were clearly built as virtualization platforms (ESXi, Hyper-V, etc.) EVE's ultimate task is to be able to run [Edge Containers](ECOS.md) as efficiently (and securely!) as possible. A running ECO is represented as a series of tasks consuming resources provided by each Edge Node (such as CPU, RAM, volumes, network instances, etc.). Tasks can further improve their performance (and security) by requesting accelerator and/or isolation services typically provided by type-1 or type-2 hypervisors and backed by hardware assisted virtualization.
+
+This mapping of Tasks to resources creates an interesting architecture where things like virtualization capabilities can be consumed transparently without the user explicitly tweaking the virtualization knobs like it would be the case with systems like libvirt/virsh.
+
+The rest of this document summarizes various implementation details we leverage for providing acceleration and isolation services to tasks.
+
+## Device models
+
+Whenever a task gets isolated into a standalone domain (either for performance or security reasons) domain gets assigned a certain, fixed amount of RAM and can have a number of virtual CPUs that would get multiplexed on top the physical CPUs available on the Edge Node. Additionally, a task may request a subset of PCI devices to be exclusively available to it (which makes them unavailable for any other domains and the host itself). Providing RAM, CPU and direct PCI assignment to a given task is one half of a job of any hypervisor. That's the easy part.
+
+The difficult (or at least much more involved part) is how to present a series of devices (network, disk, GPU, console) to the task in such a way as to allow them to be multiplexed to the actual physical devices available on the host. This is know as providing a "device model" to the domain and it is where the art of virtualization really begins.
```

Comment:
This first and second part is a bit confusing without an introduction. It might help to have an introductory paragraph, something like:

```
Assigning resources to tasks involves two types of assignment:

* dedicated, where resources are assigned directly to the workload
* virtualized, where resource are assigned in a virtualized fashion, enabling the host to multiplex them among multiple guests and the host itself
```

Then we can have those two paragraphs, in the right context. We also can shorten the paragraphs a bit.

Reply:
fixed